### DISCUSSION TITLE
Error when running Microsoft365 analysis exporting to html

### DISCUSSION THREAD

**Discussion by Bleh9000** [silverhack/monkey365#106] [General]
Hello,

When running -Instance Microsoft365 -Analysis 'Microsoft365' export to html (as well as entering tenant id, promptbehavior Selectaccount etc) I am getting a write-error stating "cannot bind argument to parameter 'items' because it is null".

I am not sure what to do here. I have tried all other commands and I get exports but when setting the analysis to 'Microsoft365' it gives me this error. When exporting with CSV it runs but it exports 2 CSVs.

o365_sway_sharing_settings.csv
o365_forms_current_user_info.csv

each CVS only has 1 row containing 8 columns but no information at all. 

The account I am selecting is my admin account which has Global Reader and Security Reader within Entra -> Roles and Administrators.

When running the command again but selecting "PRINT" I get the below information
o365_sway_sharing_settings
o365_forms_current_user_info

Am I missing something? I would like to run a m365 scan and get as much information as I can.

**Comment by silverhack** [comment]
Hey @iWarrenD94 seems a permission issue. Your user should have correct permissions to access different services. 

[https://silverhack.github.io/monkey365/permissions/](https://silverhack.github.io/monkey365/permissions/)

Hope that helps

**Reply by Bleh9000** [reply]
Hi @silverhack,

My admin account has global reader and security reader. On the Entra level. Is there a separate global reader for the m365 environment? I thought this was linked with Entra's roles?

I checked M365 admin center and I also have the global reader assigned on there.

**Reply by silverhack** [reply]
Global reader is fine excepting for SharePoint that will require SharePoint admin. In order to check for authentication issues you can append the -verbose and -informationAction continue.

**Reply by Bleh9000** [reply]
@silverhack I understand what roles are needed for sharepoint but I am mainly talking about the instance Microsoft365 Analysis 'Microsoft365' not providing much data. I am not sure what I am doing wrong.

**Reply by silverhack** [reply]
Hey @iWarrenD94,  

The best option would be to add -verbose and -informationAction continue parameter options to see what is happening under the hood. 

Regarding the Microsoft365 option, and when is included into analysis, the tool will try to get some basic information from M365 admin portal.

**Reply by Bleh9000** [reply]
is the basic information that it grabs mainly 2 csv files with only 1 row worth of information? @silverhack

**Reply by silverhack** [reply]
Hi @Bleh9000, 

That will vary depending on which license you have, but yes, that collector (Microsoft365) will try to get some basic info from Sway and Microsoft Forms. The following is an example for `o365_forms_current_user_info.json`:

```json
{
  "id": "cf939585-5fb9-4eb2-949c-3de3b7feb8d4",
  "licenses": [
    {
      "licenseName": "FORMS_PLAN_E5",
      "skus": "[\"ENTERPRISEPREMIUM\"]"
    }
  ],
  "isPaidUser": true,
  "isRestrictedUser": false,
  "userReputationTier": 1,
  "legalAgeGroup": 0,
  "language": "en-US",
  "country": ""
}
```

As you probably already know, each Microsoft 365 plan includes a number of individual services, such as SharePoint Online, Exchange Online, Microsoft Teams, etc..., and Monkey365 is mainly designed to work with M365 E3/E5 licenses. 

So if for example you want to gather data from a Microsoft 365 tenant, and you want to include all core services (Exchange, SharePoint, etc..), the following should work:

```PowerShell
 $p = @{
    Instance = 'Microsoft365';
    Analysis = @('SharePointOnline','ExchangeOnline','Purview','MicrosoftTeams','Microsoft365');
    PromptBehavior = 'SelectAccount';
    ExportTo = @('HTML','CSV');
}
Invoke-Monkey365 @p
```
And if you want to also fetch data from Microsoft Entra ID, you should add the `-IncludeEntraId` flag, as shown below:

```PowerShell
 $p = @{
    Instance = 'Microsoft365';
    Analysis = @('SharePointOnline','ExchangeOnline','Purview','MicrosoftTeams','Microsoft365');
    PromptBehavior = 'SelectAccount';
    IncludeEntraID = $true;
    ExportTo = @('HTML','CSV');
}
Invoke-Monkey365 @p
```

Hope that helps,

**Reply by Bleh9000** [reply]
Hi @silverhack
I tried running -IncludeEntraID but it took way too long. I was not able to find an explanation on what exactly that command did too.  Mind explaining the purpose of that command?

**Comment by silverhack** [comment]
So if monkey takes too long to finish is mainly because very large environments, that is, a potential Entra ID with thousand of objects (users, groups, conditional access, etc..), and you will have this and other issues with very large environments. 

For example on those environments, interactive credentials used with Monkey365 will time out and cause errors/exceptions. This is because Azure/Microsoft 365 access tokens have an expiration of roughly 60 minutes and Monkey365 does not currently support automatic refreshing tokens, and a workaround would be to execute Monkey365 with [service principal authentication](https://silverhack.github.io/monkey365/authentication/authFlows/sp/).

Another issue will be that local browsers (FF/Chrome/etc..) will crash rendering HTML reports for large environments. 

So If your long run time is caused by a large Microsoft 365 environment, there is no current workaround. If you have large tenant, you can split the collection up by services using the Analysis, IncludeEntraID and [ExcludeCollector](https://silverhack.github.io/monkey365/configuration/general-options/) flags. 

Regarding collectors used by the tool, all are located under [collectors](https://github.com/silverhack/monkey365/tree/main/collectors) folder, and an example report can be found at [here](https://silverhack.github.io/monkey365/sample/Monkey365.html).

**Comment by Bleh9000** [comment]
Hi @silverhack 
Looking at the example report you linked atr the end of your response, looking under the Microsoft 365 tab, which of those results are from the Analysis 'Microsoft365' or are those results mainly from 'SharePointOnline','ExchangeOnline','Purview','MicrosoftTeams'

**Reply by silverhack** [reply]
Hey @Bleh9000 

That example report only had issues for Exo,SP, etc.. if you are looking for Sway or Microsoft forms rules (which is part of the Microsoft365 analysis option), all of them are located into the collectors folder. 

At time of compiling that example report, no issues were found on those services.

**Reply by Bleh9000** [reply]
Hi @silverhack 
I think this may have answered our issue... What exactly is the 'Microsoft365' option scanning? or looking at? [This here](https://silverhack.github.io/monkey365/configuration/microsoft365-options/) does not provide too much information other than "Retrieves basic information from Microsoft 365". I do not think we use Microsoft sway or forms which explains why there is no information for that.

**Reply by silverhack** [reply]
Short answer: You should look for collectors with the “Microsoft365” string within the Provider attribute. 

Probably the name may cause immediate confusion about what’s included when you select “Microsoft365”. 

The Microsoft355 flag will collect information from services or basic information that are not part of core services such as Exchange, Teams, etc.. 

Internally, Monkey365 will group collectors by provider, so if you select ExchangeOnline the tool will load collectors for ExchangeOnline only if the authentication and authorisation is successful.

If you do the same with Microsoft365, the tool will load collectors tied to “Microsoft365” provider and will collect basic info from Microsoft 365 admin blade, Sway, Forms,etc.

Basically every collector file contain a metadata object with information about the collector name, id, provider, etc.. You can find an example [here](https://github.com/silverhack/monkey365/blob/301fa8e2d76e3e2729d83112905e456b12869fca/collectors/m365/MicrosoftForms/Get-MonkeyFormsUserInfo.ps1#L50)

Just to mention that I’m working in some features that will be available in future versions of the tool. One of them is a new flag that can be used to list collectors (already implemented but not tested) and flags to list rules, among others. 

Docs (and code) would benefit from further work so PR about docs, suggestions, re-wording, adding new collectors, code quality, etc, are always welcome.

Please, note that I’m the only dev for this project, and I’m doing my best in it. 

Hope this clarifies your issue.

**Reply by Bleh9000** [reply]
Hi @silverhack,
Where can I find information on everything included in the Microsoft365 collector?

**Reply by silverhack** [reply]
Hi @Bleh9000,  

Answered in last comment. 

Cheers,

**Reply by Bleh9000** [reply]
Hi @silverhack,

I see. I got confused with the collectors within the m365 folder. They looked to me as if you were to select these options 'SharePointOnline','ExchangeOnline','Purview','MicrosoftTeams' and not 'Microsoft365'
